### PR TITLE
Gates can be added to specified position in a QubitCircuit

### PR DESCRIPTION
**Pull request by abhisekupadhyaya**
This enables QubitCircuit.add_gate() to insert some gate in the middle of the circuit. This should address issue #615 .

### CODE REVIEW

**Comment by abhisekupadhyaya** [comment]
As the Cognitive Complexity exceeds the allowed limit, codeclimate is suggesting for code refactoring. Should I proceed accordingly? @nathanshammah @quantshah

**Comment by nathanshammah** [comment]
@abhisekupadhyaya, let's take it more as a suggestion. At a glance the code does not seem too convoluted.

File: qutip/qip/circuit.py [hunk 1]
```diff
@@ -220,6 +220,8 @@ def add_gate(self, gate, targets=None, controls=None, arg_value=None,
             Argument value(phi).
         arg_label: String
             Label for gate representation.
+        index : List
```
Comment (quantshah) [inline qutip/qip/circuit.py]: `list` not List. It is more about specifying the datatype than describing it.

**Comment by quantshah** [comment]
Good addition @abhisekupadhyaya. The failing test is recurring in MacOS #963 and we can ignore it now and go ahead and merge this PR @nathanshammah

**Comment by nathanshammah** [comment]
Dear @abhisekupadhyaya, 
- test run fine
- could verify the functionality
https://gist.github.com/nathanshammah/6ad41ed5c88aec5f37e10eff16866d37
add-qcircuit
Congratulations for this nice feature and thank you for your contribution. Merging

File: qutip/qip/circuit.py (hunk 1 above)
Comment (nathanshammah) [reply qutip/qip/circuit.py]: noticed this now. @abhisekupadhyaya if you can address it in a new PR, it is a quick fix.

Comment (abhisekupadhyaya) [reply qutip/qip/circuit.py]: Sure. I will get it done soon.

**Pull request merged**